pegin/getaddress.go

```diff
@@ -0,0 +1,379 @@
+		if err != nil {
+			return "", err
+		}
+		mainChainAddress = ps2h.String()
+	} else {
+		//P2WSH - P2CH
+		witnessScriptHash := sha256.Sum256(contract)
+		p2wsh, err := btcutil.NewAddressWitnessScriptHash(
+			witnessScriptHash[:],
+			btcNetwork,
+		)
+		if err != nil {
+			return "", err
+		}
+		mainChainAddress = p2wsh.String()
+	}
+
+	return mainChainAddress, nil
+}
+
+type FedpegInfo struct {
+	FedpegScript  []byte
+	FedpegProgram []byte
+}
+
+// TODO - get fedpeg details from live blockchain
+func GetFedpegInfo() (*FedpegInfo, error) {
+	return &FedpegInfo{}, nil
+}
+
+func calculateContract(
+	federationScript []byte,
+	scriptPubKey []byte,
+) ([]byte, error) {
+	contract := make([]byte, 0)
+
+	ctx, err := secp256k1.ContextCreate(secp256k1.ContextBoth)
+	if err != nil {
+		return nil, err
+	}
+
+	isLiquidV1Watchman, err := IsLiquidV1(federationScript)
+	if err != nil {
+		return nil, err
+	}
+
+	pops, err := address.ParseScript(
+		federationScript,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	liquidOpElseFound := false
+	for _, v := range pops {
+		// For liquidv1 initial watchman template, don't tweak emergency keys
+		if isLiquidV1Watchman && v.Opcode.Value == address.OP_ELSE {
+			liquidOpElseFound = true
+		}
+
+		if len(v.Data) == pubKeyLen && !liquidOpElseFound {
+			mac := hmac.New(sha256.New, v.Data)
+			mac.Write(scriptPubKey)
+			tweak := mac.Sum(nil)
+
+			_, watchman, err := secp256k1.EcPubkeyParse(ctx, v.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			_, tweaked, err := secp256k1.EcPubkeyParse(ctx, v.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			_, err = secp256k1.EcPubKeyTweakAdd(ctx, tweaked, tweak)
+			if err != nil {
+				return nil, err
+			}
+
+			_, newPub, err := secp256k1.EcPubkeySerialize(
+				ctx,
+				tweaked,
+				secp256k1.EcCompressed,
+			)
+
+			contract = append(contract, byte(len(newPub)))
+			contract = append(contract, newPub...)
+
+			// Sanity checks to reduce pegin risk. If the tweaked
+			// value flips a bit, we may lose pegin funds irretrievably.
+			// We take the tweak, derive its pubkey and check that
+			// `tweaked - watchman = tweak` to check the computation
+			// two different ways
+			_, tweaked2, err := secp256k1.EcPubkeyCreate(ctx, tweak)
+			if err != nil {
+				return nil, err
+			}
+
+			_, err = secp256k1.EcPubKeyNegate(ctx, watchman)
+			if err != nil {
+				return nil, err
+			}
+
+			vPoint := []*secp256k1.PublicKey{watchman, tweaked}
+			_, maybeTweaked2, err := secp256k1.EcPubKeyCombine(ctx, vPoint)
+			if err != nil {
+				return nil, err
+			}
+
+			_, tweaked2Bytes, err := secp256k1.EcPubkeySerialize(
+				ctx,
+				tweaked2,
+				secp256k1.EcUncompressed,
+			)
+
+			_, maybeTweaked2Bytes, err := secp256k1.EcPubkeySerialize(
+				ctx,
+				maybeTweaked2,
+				secp256k1.EcUncompressed,
+			)
+
+			if !bytes.Equal(tweaked2Bytes[:64], maybeTweaked2Bytes[:64]) {
+				return nil, errors.New("sanity check failed")
+			}
+
+		} else {
```

Comment:
Whats' the rationale here? I know is monkey porting of C version, but anything as context you can add would be appreciate.

Reply:
Dont have e2e knowledge(pegin, pegout, crypto) in order to add relevant comments in here.